Generic system config

newclarityex opened the pull request:
# Objective
Prevents duplicate implementation between IntoSystemConfigs and IntoSystemSetConfigs using a generic, adds a NodeType trait for more config flexibility (opening the door to implement https://github.com/bevyengine/bevy/issues/14195?).

## Solution
Followed writeup by @ItsDoot:
https://hackmd.io/@doot/rJeefFHc1x

Removes IntoSystemConfigs and IntoSystemSetConfigs, instead using IntoNodeConfigs with generics.

## Testing
Pending

---

## Showcase
N/A

## Migration Guide
SystemSetConfigs -> NodeConfigs<InternedSystemSet>
SystemConfigs -> NodeConfigs<ScheduleSystem>
IntoSystemSetConfigs -> IntoNodeConfigs<InternedSystemSet, M>
IntoSystemConfigs -> IntoNodeConfigs<ScheduleSystem, M>

ItsDoot commented on crates/bevy_ecs/src/schedule/config.rs:
```diff
@@ -36,24 +36,68 @@ fn ambiguous_with(graph_info: &mut GraphInfo, set: InternedSystemSet) {
 
+/// Stores data to differentiate different Node types
+pub trait NodeType {
+    /// Used in `NodeConfig`
```
```suggestion
    /// Additional data used to configure a node. Stored in [`NodeConfig`].
```

We could elaborate here a bit.

BTW, (for bevy at least) 99% of the time you want to wrap anything quoted in ticks with square brackets so rustdoc picks them up.

ItsDoot commented on crates/bevy_ecs/src/schedule/config.rs:
```diff
@@ -36,24 +36,68 @@ fn ambiguous_with(graph_info: &mut GraphInfo, set: InternedSystemSet) {
+pub trait NodeType {
+    /// Used in `NodeConfig`
+    type Metadata;
+    /// Used in `NodeConfigs`
```
```suggestion
    /// Additional data used to configure a group of nodes. Stored in [`NodeConfigs`].
```
Same as above.

ItsDoot commented on crates/bevy_ecs/src/schedule/config.rs:
```diff
@@ -36,24 +36,68 @@ fn ambiguous_with(graph_info: &mut GraphInfo, set: InternedSystemSet) {
+    /// Used in `NodeConfigs`
+    type GroupMetadata;
+
+    /// Initialize `NodeConfig` from `NodeType`
```
```suggestion
    /// Initializes a configuration from this node.
```
We can use normal english here since `NodeConfig` is already given as the return type, and this is a trait method on `Nodetype`.

ItsDoot commented on crates/bevy_ecs/src/schedule/config.rs:
```diff
@@ -36,24 +36,68 @@ fn ambiguous_with(graph_info: &mut GraphInfo, set: InternedSystemSet) {
+    type GroupMetadata;
+
+    /// Initialize `NodeConfig` from `NodeType`
+    fn config(self) -> NodeConfig<Self>
```
```suggestion
    fn into_config(self) -> NodeConfig<Self>
```
This change follows rust naming conventions for functions that take `self`.

ItsDoot commented on crates/bevy_ecs/src/schedule/config.rs:
```diff
@@ -36,24 +36,68 @@ fn ambiguous_with(graph_info: &mut GraphInfo, set: InternedSystemSet) {
+    type GroupMetadata = Chain;
+
+    fn config(self) -> NodeConfig<Self> {
+        let sets = self.default_system_sets().into_iter().collect();
```
```suggestion
        let sets = self.default_system_sets().clone();
```
This is actually something I noticed a while ago but haven't gotten around to changing, lol.

ItsDoot commented on crates/bevy_ecs/src/schedule/config.rs:
```diff
@@ -62,35 +106,19 @@ pub enum NodeConfigs<T> {
         collective_conditions: Vec<BoxedCondition>,
+        /// Metadata for `NodeConfigs`
+        metadata: T::GroupMetadata,
         /// See [`Chain`] for usage.
         chained: Chain,
```
```suggestion
```
This can be removed because it's provided by the `metadata: T::GroupMetadata` (recall `type GroupMetadata = Chain;`).

ItsDoot commented on crates/bevy_ecs/src/schedule/config.rs:
```diff
@@ -62,35 +106,19 @@ pub enum NodeConfigs<T> {
         configs: Vec<NodeConfigs<T>>,
         /// Run conditions applied to everything in the tuple.
         collective_conditions: Vec<BoxedCondition>,
+        /// Metadata for `NodeConfigs`
```
```suggestion
        /// Metadata to be applied to all elements in the tuple.
```
We can be a little more descriptive.

ItsDoot commented on crates/bevy_ecs/src/schedule/config.rs:
```diff
@@ -450,12 +505,16 @@ where
 
-impl IntoSystemConfigs<()> for SystemConfigs {
+impl<T: NodeType<Metadata = GraphInfo>> IntoNodeConfigs<T, ()> for NodeConfigs<T> {
+    // impl<T: NodeType<Metadata = GraphInfo>, Marker> IntoNodeConfigs<T, Marker> for NodeConfigs<T> {
```
```suggestion
```
Left over comment?

ItsDoot commented on crates/bevy_ecs/src/schedule/config.rs:
```diff
@@ -496,12 +555,18 @@ impl IntoSystemConfigs<()> for SystemConfigs {
     }
 
-    fn distributive_run_if<M>(mut self, condition: impl Condition<M> + Clone) -> SystemConfigs {
+    fn distributive_run_if<M>(mut self, condition: impl Condition<M> + Clone) -> NodeConfigs<T>
+    where
+        T: NodeType,
+    {
```
```suggestion
    fn distributive_run_if<M>(mut self, condition: impl Condition<M> + Clone) -> NodeConfigs<T> {
```
This is a redundant bound since the trait already specifies `T: NodeType`.

ItsDoot commented on crates/bevy_ecs/src/schedule/config.rs:
```diff
@@ -496,12 +555,18 @@ impl IntoSystemConfigs<()> for SystemConfigs {
     }
 
-    fn run_if<M>(mut self, condition: impl Condition<M>) -> SystemConfigs {
+    fn run_if<M>(mut self, condition: impl Condition<M>) -> NodeConfigs<T>
+    where
+        T: NodeType,
+    {
```
```suggestion
    fn run_if<M>(mut self, condition: impl Condition<M>) -> NodeConfigs<T> {
```
Same as above.

ItsDoot commented on crates/bevy_ecs/src/schedule/schedule.rs:
```diff
@@ -372,7 +375,10 @@ impl Schedule {
     /// Configures a collection of system sets in this schedule, adding them if they does not exist.
     #[track_caller]
-    pub fn configure_sets(&mut self, sets: impl IntoSystemSetConfigs) -> &mut Self {
+    pub fn configure_sets<Marker>(
+        &mut self,
+        sets: impl IntoNodeConfigs<InternedSystemSet, Marker>,
+    ) -> &mut Self {
```
```suggestion
    pub fn configure_sets<M>(
        &mut self,
        sets: impl IntoNodeConfigs<InternedSystemSet, M>,
    ) -> &mut Self {
```
For functions that use the marker generic, we use `M` instead of the full `Marker` name.

ItsDoot commented on crates/bevy_ecs/src/schedule/schedule.rs:
```diff
@@ -950,15 +960,18 @@ impl ScheduleGraph {
 
     #[track_caller]
-    fn configure_sets(&mut self, sets: impl IntoSystemSetConfigs) {
+    fn configure_sets<Marker>(&mut self, sets: impl IntoNodeConfigs<InternedSystemSet, Marker>) {
```
```suggestion
    fn configure_sets<M>(&mut self, sets: impl IntoNodeConfigs<InternedSystemSet, M>) {
```
Same deal as above.

ItsDoot commented on crates/bevy_ecs/src/schedule/schedule.rs:
```diff
@@ -1537,7 +1550,9 @@ struct ProcessConfigsResult {
 trait ProcessNodeConfig: Sized {
     /// Process a single [`NodeConfig`].
-    fn process_config(schedule_graph: &mut ScheduleGraph, config: NodeConfig<Self>) -> NodeId;
+    fn process_config(schedule_graph: &mut ScheduleGraph, config: NodeConfig<Self>) -> NodeId
+    where
+        Self: NodeType;
```
Since `ProcessNodeConfig` only has this single function, the `NodeType` bound should be instead moved to the trait definition like so:

```rust
trait ProcessNodeConfig: NodeType + Sized {
```

I can't directly suggest it unfortunately because GitHub won't let me.

ItsDoot reviewed (changes requested):
First pass done! Good work so far!

However, one thing I don't mention below is that the `impl_system_collection` and `impl_system_set_collection` macros that provide implementations for tuples should be merged into a single `impl_node_type_collection` macro that uses a `T: NodeType` bound instead of hardcoding to `ScheduleSystem` or `InternedSystemSet`. It should be as simple as taking one of the macros, renaming it, adding the bounds, and using `T` in place of either of the aforementioned types.

hymm commented:
`NodeConfigs` feels a little too generic. Would it make sense to call it `ScheduleConfigs`?

newclarityex commented:
> `NodeConfigs` feels a little too generic. Would it make sense to call it `ScheduleConfigs`?

Yup, updated NodeConfigs and NodeConfig to ScheduleConfigs, although kept NodeType as is.

alice-i-cecile commented on crates/bevy_ecs/src/schedule/config.rs:
```diff
@@ -36,61 +36,87 @@ fn ambiguous_with(graph_info: &mut GraphInfo, set: InternedSystemSet) {
     }
 }
 
+/// Stores data to differentiate different Node types
```
Needs doc links to `Node`.

alice-i-cecile commented on crates/bevy_ecs/src/schedule/config.rs:
```diff
@@ -36,61 +36,87 @@ fn ambiguous_with(graph_info: &mut GraphInfo, set: InternedSystemSet) {
-pub enum NodeConfigs<T> {
+/// A collections of generic [`ScheduleConfig`]s.
+pub enum ScheduleConfigs<T: NodeType> {
     /// Configuration for a single node.
```
Is this doc comment still true?

alice-i-cecile reviewed (changes requested):
I really like the net effect here, but IMO the `NodeType` name is quite confusing, and not a good fit for a trait. How would `Schedulable` work for you?

Some little doc nits too :)

newclarityex commented:
> I really like the net effect here, but IMO the `NodeType` name is quite confusing, and not a good fit for a trait. How would `Schedulable` work for you?
> 
> Some little doc nits too :)

Renamed and updated some docs :>

newclarityex commented:
> Manually resolved merge conflicts. You might need a cargo fmt in there. Ping me to re-attempt the merge if needed.

i think we can reattempt it now

alice-i-cecile commented:
Awesome, thanks :) I appreciate it: merge conflicts can be super annoying on a project this big.

mnmaita commented:
FYI the migration guide doesn't seem to be up to date based on some naming changes made later.

newclarityex commented:
> FYI the migration guide doesn't seem to be up to date based on some naming changes made later.


Updated
